Added DWD WarnApp Sensor

runningman84 opened the pull request:
## Description:
This sensor adds support for current and advance weather warnings from DWD (Deutsche Wetter Dienst) for Germany. This sensor uses the same data as the WarnWetter Android/IOS App.

The warning level is between 0 (no danger) and 4:
* Warnungen vor extremem Unwetter (Stufe 4)
* Unwetterwarnungen (Stufe 3)
* Warnungen vor markantem Wetter (Stufe 2)
* Wetterwarnungen (Stufe 1)

It can be a bit tricky to find out the correct region name because the source json files only shows a region if a current or advance warning is there:
https://www.dwd.de/DWD/warnungen/warnapp_landkreise/json/warnings.json?jsonp=loadWarnings
The region_name should match the name of the regions listed on this page:
https://www.dwd.de/DE/wetter/warnungen_landkreise/warnWetter_node.html?ort=Hamburg

**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io) with documentation (if applicable):** home-assistant/home-assistant.github.io#3062

## Example entry for `configuration.yaml` (if applicable):
```yaml
sensor:
  - platform: dwd_weather_warnings
    region_name: 'Hansestadt Hamburg'   # optional
```
Here is a screenshot:
https://community.home-assistant.io/t/dwd-warnapp-sensor-amtliche-warnungen-des-deutschen-wetterdienstes/22699/1

## Checklist:

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [ ] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [x] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [ ] New dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [x] New files were added to `.coveragerc`.

If the code does not interact with devices:
  - [ ] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [ ] Tests have been added to verify that the new code works.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

runningman84 committed:
Added DWD WarnApp Sensor

homeassistant commented:
Hi @runningman84, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#8657).

Once you do that we will be able to review and accept this pull request.

Thanks!

runningman84 committed:
Fixed some idents and spaces

runningman84 commented on homeassistant/components/sensor/dwd_warnapp.py:
```diff
@@ -0,0 +1,245 @@
+import homeassistant.util.dt as dt_util
+
+_LOGGER = logging.getLogger(__name__)
+_ENDPOINT = '/DWD/warnungen/warnapp_landkreise/json/warnings.json?jsonp=loadWarnings'
```
how ca we fix this one? I cannot change the corresponding url, splitting it into multiple lines seems odd to me.

balloob replied:
Fix it by not hardcoding the url arguments in the endpoint.

runningman84 committed:
Removed unused imports

runningman84 committed:
Removed comment

balloob commented on homeassistant/components/sensor/dwd_warnapp.py:
```diff
@@ -0,0 +1,242 @@
+DEFAULT_METHOD = 'GET'
+DEFAULT_NAME = 'DWD-Warnapp'
+DEFAULT_SSL = True
+DEFAULT_VERIFY_SSL = True
```
Why did you implement this option? Doesn't it default fetch `dwd.de` which I assume has SSL working?

balloob commented on homeassistant/components/sensor/dwd_warnapp.py:
```diff
@@ -0,0 +1,240 @@
+        self.available = True
+        self.update()
+
+    #@Throttle(SCAN_INTERVAL)
```
You will need to implement throttle or you're hammering their API

balloob commented on homeassistant/components/sensor/dwd_warnapp.py:
```diff
@@ -0,0 +1,242 @@
+        """Initialize the data object."""
+        from homeassistant.components.sensor.rest import RestData
+
+        uri_scheme = 'https://' if use_ssl else 'http://'
```
Why is this an option? Isn't the website always the same

balloob reviewed (changes requested):
There are a lot of options in here that I don't really see why they would make sense for querying a web service (verify SSL, use SSL etc)

runningman84 committed:
Some fixes

runningman84 committed:
Added throttle

runningman84 committed:
Renamed sensor to dwd weather warnings

runningman84 commented:
@balloob I added the throttle function and renamed it to dwd weather warnings. 
I do not have a good idea how to shorten the lines to be below 80 chars. Is that a hard limit? If yes please provide some advice how to do this...

runningman84 committed:
Renamed test file

balloob commented:
```python
 if (event['level'] >=
         data['{}_warning_level'.format(mykey)]):
```

balloob commented:
Yes, it is a hard limit.

runningman84 committed:
shorten lines

runningman84 committed:
shorten lines

fabaff commented on homeassistant/components/sensor/dwd_weather_warnings.py:
```diff
@@ -0,0 +1,242 @@
+
+_LOGGER = logging.getLogger(__name__)
+
+DEFAULT_NAME = 'DWD-Wetter-Warnungen'
```
Should be in English. If the user can set `name:` if he/she want it in German.

runningman84 replied:
ok

fabaff commented on homeassistant/components/sensor/dwd_weather_warnings.py:
```diff
@@ -0,0 +1,242 @@
+DEFAULT_NAME = 'DWD-Wetter-Warnungen'
+
+CONF_REGION_NAME = 'region_name'
+DEFAULT_REGION_NAME = 'Hansestadt Hamburg'
```
This default does only make sense for a small group of users.

runningman84 replied:
I will remove the default region

fabaff commented on homeassistant/components/sensor/dwd_weather_warnings.py:
```diff
@@ -0,0 +1,242 @@
+
+    # pylint: disable=no-member
+    @property
+    def device_state_attributes(self):
```
Can you please add an attribution?

runningman84 replied:
what do you mean?

fabaff commented on homeassistant/components/sensor/dwd_weather_warnings.py:
```diff
@@ -0,0 +1,242 @@
+
+    def __init__(self, hass, api, name, variable):
+        """Initialize a DWD-Weather-Warnings sensor."""
+        self._hass = hass
```
You doesn't seems to use `self._hass`.

runningman84 replied:
I will remove it

fabaff commented on homeassistant/components/sensor/dwd_weather_warnings.py:
```diff
@@ -0,0 +1,242 @@
+
+    def __init__(self, region_name):
+        """Initialize the data object."""
+        from homeassistant.components.sensor.rest import RestData
```
Please place this in the file header.

runningman84 replied:
ok

runningman84 committed:
Implemented changes requested by fabaff

runningman84 committed:
added ATTRIBUTION

runningman84 committed:
move ATTRIBUTION to existing method

runningman84 commented:
@fabaff can you accept the pull request or is still something missing?

runningman84 committed:
fixed lint tests

fabaff commented:
There are still lint issues.

runningman84 committed:
Fix linter issues

runningman84 committed:
Fix linter issues

runningman84 committed:
Fix linter

syssi commented on homeassistant/components/sensor/dwd_weather_warnings.py:
```diff
@@ -0,0 +1,242 @@
+        elif self._var_id == 'advance_warning_level':
+            prefix = 'advance'
+
+        data['warning_count'] = self._api.data[prefix + '_warning_count']
```
Local variable 'prefix' might be referenced before assignment. Compare L122, L124.

runningman84 replied:
ok I will add a else path and throw an exception there...

syssi commented on homeassistant/components/sensor/dwd_weather_warnings.py:
```diff
@@ -0,0 +1,243 @@
+
+    def __init__(self, region_name):
+        """Initialize the data object."""
+
```
Just remove this blank line.

runningman84 replied:
ok

syssi commented on homeassistant/components/sensor/dwd_weather_warnings.py:
```diff
@@ -0,0 +1,243 @@
+            json_string = self._rest.data[24:len(self._rest.data) - 2]
+            json_obj = json.loads(json_string)
+
+            data = {}
```
You could save a line here. My IDE complains about. ;-)

- data = {}
-
- data['time'] = json_obj['time']
+ data = {'time': json_obj['time']}

runningman84 replied:
good catch

syssi commented on homeassistant/components/sensor/dwd_weather_warnings.py:
```diff
@@ -0,0 +1,243 @@
+    @property
+    def device_state_attributes(self):
+        """Return the state attributes of the DWD-Weather-Warnings."""
+        data = {}
```
Same here:

```
        data = {
            ATTR_ATTRIBUTION: ATTRIBUTION,
            'region_name': self._api.region_name,
        }
```

runningman84 replied:
good catch

syssi commented:
Good job! :+1:

runningman84 committed:
Fixed linter

runningman84 commented:
@fabaff and @balloob all checks have passed

balloob commented:
Thanks! 🐬
